### ISSUE TITLE
[RFC] A sugary `defineConfig()` function annotation for pollen's config file

### ISSUE THREAD

**Issue by aldy505** [heybokeh/pollen#55]
Per this issue comment (https://github.com/heybokeh/pollen/issues/46#issuecomment-968052304) and this one (https://github.com/heybokeh/pollen/issues/47#issuecomment-968057497), I'm glad that there will be a `pollen.config.js` to extend pollen. I'm just wondering that it might be a good idea to add a little sugar to the config file without forcing the non-typescript users (at this point of writing, I don't know whether pollen's config file supports the `.ts` extension or not) to use JSdoc annotation for `/** @type import("pollen.Config") */`.

I'm suggesting something like Vite's ([link](https://github.com/vitejs/vite/blob/4ece719229ac606aac93e5a267228582f54574bc/packages/vite/src/node/config.ts#L64-L72)) and WindiCSS' ([link](https://github.com/windicss/windicss/blob/main/src/helpers/config.ts)) config file.

So, it would be something like:

```js
/** pollen.config.js */
import pollen, { defineConfig } from "pollen";

export default defineConfig({
  output: 'styles/pollen.css',
  modules: {
    font: { sans: `"Inter", sans-serif` },
    letter: { ...pollen.letter, sm: '-0.02em' },
    radius: { ...pollen.radius, xs: '4px ' },
    grid: { ...pollen.grid, pageGutter: 'var(--grid-gap)' },
    color: {
      grey100: '#f2f4f7',
	  ...
    }
  }
})
```

Or, if `import pollen from "pollen"` sounds like a no for you, because you might need to export some other things on the `pollen.js` file, we could make the exported configs on a separate file (this wouldn't cause any breaking change tho).

```js
import { defineConfig } from "pollen/helpers";
import { letter } from "pollen/typography";
import { radius } from "pollen/ui";
import { grid } from "pollen/grid";

export default defineConfig({
  output: 'styles/pollen.css',
  modules: {
    font: { sans: `"Inter", sans-serif` },
    letter: { ...letter, sm: '-0.02em' },
    radius: { ...radius, xs: '4px ' },
    grid: { ...grid, pageGutter: 'var(--grid-gap)' },
    color: {
      grey100: '#f2f4f7',
	  ...
    }
  }
})
```

or if you want to make it a bit neat and put everything on the entry point which you'll only need to do `export * from "./typography"`...

```js
import { defineConfig, defaults } from "pollen";

export default defineConfig({
  output: 'styles/pollen.css',
  modules: {
    font: { sans: `"Inter", sans-serif` },
    letter: { ...defaults.letter, sm: '-0.02em' },
    radius: { ...defaults.radius, xs: '4px ' },
    grid: { ...defaults.grid, pageGutter: 'var(--grid-gap)' },
    color: {
      grey100: '#f2f4f7',
	  ...
    }
  }
})
```

If there's another feedback, I'd love to hear it. I might get to work on the PR as soon as I can :)

**Comment by madeleineostoja** [comment]
Hey, thanks for the ideas! I don’t use vite or windicss so I’m not across their config implementation, but how is this different to providing the `pollen` defaults in an argument to the config function (how it’s currently done) vs an import? Imo having it as an argument in the config function is quite ergonomic. But very open to arguments on the contrary! 

In terms of the actual `defineConfig` function, I suppose it is more explicit than hijacking a default module export, but again the latter has good precedent with webpack, rollup, etc. and it’s what came to mind instinctively when thinking of a typical config file.

I’m still documenting the config system but here’s an example of config with the current setup: https://github.com/heybokeh/pollen/issues/47#issuecomment-968057497

Oh and yeah the default import from the `pollen-css` package should always be the precompiled default css bundle imo. But if we wanted to put the defaults in an import could very easily throw it in a submodule like `pollen-css/defaults`.

**Comment by aldy505** [comment]
The `defineConfig` on Vite and WindiCSS basically just provides a type annotation, because it's a typed object right? On Vite and WindiCSS I usually prefer to use `defineConfig` but they also allow the direct use of `export default { ViteSettings... }`.

I know it's not so much of a different other than the type annotation, which I also know if you're using JSDoc, you *might* also get the type annotation there too. But I think it'd be a lot cleaner to use `defineConfig`. Nevertheless, I guess it would be very cool to support both (the `defineConfig` and `export default (pollen) => ...`.

Oh anyway, I'm writing this because I'm currently using Pollen for a new project (using Svelte), but I need some more variables, in which I don't want to make the `:root` dirty tho haha! Seeing that it got a config files makes me a lot happy, I think it'll be a lot cleaner.

**Comment by madeleineostoja** [comment]
Gotcha. So there’s two seperate proposals here:

1. Export a `defineConfig` function for better typing.
2. Provide defaults as seperate import rather than argument to config function, and make config a plain object export 

(1) would be a non-breaking change we can add at any time. I’ll look into it. I dunno how well it’ll work in a js file compared to annotation? Has to be a JS file currently since it’s a node cli. Don’t really wanna add a babel runtime or something to support ts config files 

(2) would be a breaking change. It’s not a pattern I’ve seen in config files before, but I do like how explicit it is. Currently I feel like having the defaults as an arg to a config function is more intuitive, which is already more explicit than automatically merging config objects with silent defaults (like tailwind extend keys in their config)

**Comment by madeleineostoja** [comment]
Opened a new issue for the second point for v4, this can remain open as something to explore seperately

**Comment by madeleineostoja** [comment]
`defineConfig` helper added in `v4.1.0`

Usage: 

```js
const { defineConfig } = require('pollen-css');

// Without extending Pollen defaults
module.exports = defineConfig({ ... });

// With extending Pollen defaults
module.exports = defineConfig(pollen => ({ ... }));
```

**Comment by madeleineostoja** [comment]
Included in `v4.1.1` 🎉

Usage: 

```js
const { defineConfig } = require('pollen-css/utils');

// Overwriting defaults
module.exports = defineConfig({ ... });

// Extending defaults
module.exports = defineConfig(pollen => ({ ... }));
```
